### PR TITLE
18526 - import Date picker components

### PR DESCRIPTION
**Pull request by ketaki-deodhar**
*Issue #:* /bcgov/entity#18526

*Description of changes:*

- Create new components RangeDatePicker, InputFieldDatePicker, BaseDatePicker
- update libraries to make unit tests work

BaseDatePicker: 

image

InputFieldDatePicker: 

https://github.com/bcgov/bcrs-shared-components/assets/116035339/4a3e7e66-fe45-4164-993e-baa38774ce04

RangeDatePicker:

image

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the business-filings-ui license (Apache 2.0).

### CODE REVIEW

File: tests/unit/DetailComment.spec.ts
Comment (ketaki-deodhar) [inline tests/unit/DetailComment.spec.ts:76]: Couldn't get this working but will keep looking for solution.......

File: package.json
Comment (ketaki-deodhar) [inline package.json:46]: needed to install `@vue/test-utils@next` to get compatible version of `vue-test-utils` with vue 3

File: src/components/base-date-picker/package.json
Comment (severinbeauvais) [reply src/components/base-date-picker/package.json:3]: No, leave as 0.0.0 for your PR. When I publish this, it will get set to 1.0.0.

File: package.json
Comment (severinbeauvais) [inline package.json:49]: What's this for?

File: package.json
Comment (severinbeauvais) [inline package.json:68]: What's this for?

File: src/components/base-date-picker/BaseDatePicker.vue
Comment (severinbeauvais) [inline src/components/base-date-picker/BaseDatePicker.vue:1]: Did you change anything in here or is it the same as what we got from Cameron?

File: src/components/input-field-date-picker/InputFieldDatePicker.vue
Comment (severinbeauvais) [inline src/components/input-field-date-picker/InputFieldDatePicker.vue:1]: This looks similar to our original date picker. Does it fully replace it?

File: src/components/range-date-picker/package.json
Comment (severinbeauvais) [inline src/components/range-date-picker/package.json:11]: Looks like it doesn't depend on  lodash or vue-facing-decorator.

Looks like it depends on:
- vue
- shared base-date-picker

File: tests/unit/DetailComment.spec.ts
Comment (severinbeauvais) [inline tests/unit/DetailComment.spec.ts:44]: Does this test need more fixing or is it good now?

File: tests/unit/DetailComment.spec.ts
Comment (severinbeauvais) [reply tests/unit/DetailComment.spec.ts:76]: You may have to trigger an update on the textarea (submit or blur maybe?), to get it to recognize the new value you set. Do a global search (maybe in other UIs, too) for `setValue` and see what was done.

File: package.json
Comment (ketaki-deodhar) [reply package.json:68]: This is to import vitest-axe/extend-expect in test setup file to make sure tests work with Vitest

Ref: https://github.com/chaance/vitest-axe

File: package.json
Comment (ketaki-deodhar) [reply package.json:49]: Something I added for resolving an issue with imports but I think this is not useful and will remove it.

File: src/components/base-date-picker/BaseDatePicker.vue
Comment (ketaki-deodhar) [reply src/components/base-date-picker/BaseDatePicker.vue:1]: Pretty much the same. Only change was to use correct css for validation message.

File: src/components/input-field-date-picker/InputFieldDatePicker.vue
Comment (ketaki-deodhar) [reply src/components/input-field-date-picker/InputFieldDatePicker.vue:1]: Almost same. There are slight differences.
Main branch/DEV:

image

Label
image

Vue3 branch:

Style is a bit different
image

Does not have label
image

File: src/components/input-field-date-picker/InputFieldDatePicker.vue
Comment (ketaki-deodhar) [reply src/components/input-field-date-picker/InputFieldDatePicker.vue:161]: This is in DEV/Main branch:
image

The wrapper was wider and the base date picker width was smaller. This is fixed now to look same as above.
image

Do you think it should be smaller?

**Comment by ketaki-deodhar** [comment]
> Check with Cam on whether he has any date component unit tests you could add here.

I checked with Cameron and unfortunately they do not have any unit tests. Does it make sense to create a new ticket for building unit tests?

File: src/components/input-field-date-picker/InputFieldDatePicker.vue
Comment (severinbeauvais) [reply src/components/input-field-date-picker/InputFieldDatePicker.vue:161]: UI/UX will have final say on size. I was commenting on this (below) but you've fixed it now :)

image

File: src/components/input-field-date-picker/InputFieldDatePicker.vue
Comment (severinbeauvais) [reply src/components/input-field-date-picker/InputFieldDatePicker.vue:1]: @yuisotozaki Please see differences above between our Vue2 date picker component and the Vue3 one. (Underneath, they're completing different code.) Should we spend any time making the Vue3 one look the same or can we live with the new one as-is?

Comment (yuisotozaki) [reply src/components/input-field-date-picker/InputFieldDatePicker.vue:1]: @severinbeauvais The plan is to try to not customize as much as we can. We'll want to unify the primary colour of the buttons in Vue3 date picker popup. As for the label that tuck away, I'll check with the design team today. It's a bad timing right now as we're waiting on GDX design system release and Figma licensing. FYI: Alexis Ronse is our point person for all design system things going forward.

Comment (severinbeauvais) [reply src/components/input-field-date-picker/InputFieldDatePicker.vue:1]: What I'm hearing is that we won't change the Vue3 data picker above right now.

Note that we have (or will shortly have) two date pickers to choose from: the Vuetify date picker wrapper (same thing we did in Vue2) and the custom-built one from Assets team (as above). Ultimately we should winnow that down to 1 but that's a future concern.

Comment (leodube-aot) [reply src/components/input-field-date-picker/InputFieldDatePicker.vue:1]: Just to add to Sev's comment, my current ticket (bcgov/entity#18779) is for updating our existing Vue 2 DatePicker component to Vue 3, so probably no need to try to get BaseDatePicker to look like our existing one.

Comment (severinbeauvais) [reply src/components/input-field-date-picker/InputFieldDatePicker.vue:1]: Leo, it could go either way: either we make yours look like BaseDatePicker, or we make BaseDatePicker looks like yours, or we leave them different.

One reason for leaving BaseDatePicker alone is to keep the code as we got it. This would allow Assets team to use this shared component with no changes.

Maybe the question for @yuisotozaki is, "which date picker should we use when we update our UIs to Vue3 -- the Vuetify one or the custom one?"

Comment (severinbeauvais) [reply src/components/input-field-date-picker/InputFieldDatePicker.vue:1]: Personally, I'd prefer to use the Vuetify one (assuming it's compatible), as it's less code for us to manage. Though it is more risky, and we do have the source code for the custom one...

File: tests/unit/DetailComment.spec.ts
Comment (ketaki-deodhar) [reply tests/unit/DetailComment.spec.ts:77]: I haven't been able to resolve this and looking into it. I would like to merge this PR first so that other developers can start updating unit tests for the components they are working on. I will keep a note of this and work on this unit test later .

Comment (leodube-aot) [reply tests/unit/DetailComment.spec.ts:77]: @ketaki-deodhar Not sure if you've found this already, but this comment from filings UI might help. Apparently you can't set a textarea's value so you might have to do it explicitly: https://github.com/bcgov/business-filings-ui/blob/main/tests/unit/DetailComment.spec.ts#L75-L82

**Review by leodube-aot — APPROVED** [review]
LGTM

File: tests/unit/DetailComment.spec.ts
Comment (severinbeauvais) [reply tests/unit/DetailComment.spec.ts:77]: Sounds good to me.

Don't spend too long (don't get stuck) on anything. For now we just need to make things like-for-like in Vue3. Thanks.

File: tests/unit/DetailComment.spec.ts
Comment (severinbeauvais) [inline tests/unit/DetailComment.spec.ts:51]: This isn't the same as the previous test.

Previously, we started with no props and then set them. Now, you start with initial props.

I'd prefer to see the previous unit test left alone, skipped if you can't get it working. Add a new test if needed.

Also, why did you remove old line 65?

File: tests/unit/DetailComment.spec.ts
Comment (severinbeauvais) [inline tests/unit/DetailComment.spec.ts:67]: Same comment as above.

File: src/components/base-date-picker/package.json
Comment (severinbeauvais) [reply src/components/base-date-picker/package.json:3]: Fixed, resolving.

**Review by severinbeauvais — APPROVED** [review]
Figure out what you're going to do with the tests and then this is good to merge.

File: tests/unit/DetailComment.spec.ts
Comment (severinbeauvais) [inline tests/unit/DetailComment.spec.ts:88]: Is there a good reason for removing this? It was initially added to prevent a memory leak, but maybe is no longer needed.

File: tests/unit/DetailComment.spec.ts
Comment (ketaki-deodhar) [reply tests/unit/DetailComment.spec.ts:51]: For this unit test, earlier there was a line `wrapper.setProps({ value: 'testing 1 2 3' })` which I just changed to setting prop while mounting.  I will revert it back and skip the test.

File: tests/unit/DetailComment.spec.ts
Comment (ketaki-deodhar) [reply tests/unit/DetailComment.spec.ts:88]: Actually destroy() is now changed to unmount(). Changed it accordingly.
Ref: https://test-utils.vuejs.org/migration/#destroy-is-now-unmount-to-match-Vue-3

File: tests/unit/DetailComment.spec.ts
Comment (severinbeauvais) [reply tests/unit/DetailComment.spec.ts:53]: Sure.

"setProps" is different from mounting the component with props. In the former, you are _changing_ the props. In the latter, they are initialized with those values. The former can cause getters and watchers to react.

Comment (severinbeauvais) [reply tests/unit/DetailComment.spec.ts:53]: Thanks for the new test (in addition to the original, skipped test).

FWIW, I consider this conversation resolved.

File: tests/unit/DetailComment.spec.ts
Comment (severinbeauvais) [reply tests/unit/DetailComment.spec.ts:88]: 🤯 

I guess this will have to be changed everywhere. @leodube-aot @PaulGarewal @JazzarKarim, FYI.

**Pull request merged**
